relay-sampling/src/evaluation.rs

```diff
@@ -95,139 +114,193 @@ pub fn merge_configs_and_match(
+                matched_rule_ids.push(rule.id);
 
-                    if rule.ty == RuleType::Trace {
-                        if let Some(dsc) = dsc {
-                            seed = Some(dsc.trace_id);
-                        }
+                if rule.ty == RuleType::Trace {
+                    if let Some(dsc) = dsc {
+                        seed = Some(dsc.trace_id);
                     }
+                }
 
-                    let value = evaluator.evaluate(now);
-                    match rule.sampling_value {
-                        SamplingValue::Factor { .. } => accumulated_factors *= value,
-                        SamplingValue::SampleRate { .. } => {
-                            return Some(SamplingMatch {
-                                sample_rate: (value * accumulated_factors).clamp(0.0, 1.0),
-                                seed: match seed {
-                                    Some(seed) => seed,
-                                    // In case we are not able to generate a seed, we will return a no
-                                    // match.
-                                    None => return None,
-                                },
-                                matched_rule_ids: MatchedRuleIds(matched_rule_ids),
-                            });
-                        }
+                match rule.sampling_value {
+                    SamplingValue::Factor { .. } => accumulated_factors *= value,
+                    SamplingValue::SampleRate { .. } => {
+                        let sample_rate = {
+                            let base = (value * accumulated_factors).clamp(0.0, 1.0);
+                            if adjust_sample_rate {
+                                dsc.and_then(|d| d.sample_rate)
+                                    .map(|client_sample_rate| {
+                                        DynamicSamplingContext::adjust_sample_rate(
+                                            client_sample_rate,
+                                            base,
+                                        )
+                                    })
+                                    .unwrap_or(base)
+                            } else {
+                                base
+                            }
+                        };
+
+                        let Some(seed) = seed else {
+                            return SamplingResult::NoMatch;
+                        };
+
+                        let is_kept = sampling_match(sample_rate, seed);
+
+                        return SamplingResult::Match {
+                            sample_rate,
+                            seed,
+                            matched_rules: MatchedRuleIds(matched_rule_ids),
+                            is_kept,
+                        };
                     }
                 }
             }
         }
-
-        // In case no match is available, we won't return any specification.
-        None
     }
+
```

Comment:
question: if there are no `SamplingValue::SampleRate` in the rules, should we log that?